core/src/raw/adapters/kv/backend.rs

```diff
@@ -228,87 +226,52 @@ impl oio::BlockingList for KvLister {
 pub struct KvWriter<S> {
     kv: Arc<S>,
     path: String,
-
-    buffer: FlexBuffer,
+    buffer: QueueBuf,
 }
 
 impl<S> KvWriter<S> {
     fn new(kv: Arc<S>, path: String) -> Self {
         KvWriter {
             kv,
             path,
-            buffer: FlexBuffer::Active(BytesMut::new()),
+            buffer: QueueBuf::new(),
         }
     }
 }
 
-/// TODO: replace with oio::FlexBuf.
-enum FlexBuffer {
-    Active(BytesMut),
-    Frozen(Bytes),
-}
-
 /// # Safety
 ///
 /// We will only take `&mut Self` reference for KvWriter.
 unsafe impl<S: Adapter> Sync for KvWriter<S> {}
 
 impl<S: Adapter> oio::Write for KvWriter<S> {
     async fn write(&mut self, bs: Buffer) -> Result<usize> {
-        match &mut self.buffer {
-            FlexBuffer::Active(buf) => {
-                buf.extend_from_slice(bs.chunk());
-                Ok(bs.chunk().len())
-            }
-            FlexBuffer::Frozen(_) => {
-                unreachable!("KvWriter should not be frozen during poll_write")
-            }
-        }
+        let ret = bs.len();
+        self.buffer.push(bs);
+        Ok(ret)
     }
 
     async fn close(&mut self) -> Result<()> {
-        let buf = match &mut self.buffer {
-            FlexBuffer::Active(buf) => {
-                let buf = buf.split().freeze();
-                self.buffer = FlexBuffer::Frozen(buf.clone());
-                buf
-            }
-            FlexBuffer::Frozen(buf) => buf.clone(),
-        };
-
-        self.kv.set(&self.path, &buf).await
+        let buf = self.buffer.collect();
```

Comment:
I wonder if we can avoid clone here but set `self.buffer = QueueBuf::default()` with `mem::replace`. And then call `into_buffer` as introduced in #4497.